feature-wallet-impl/src/main/java/io/novafoundation/nova/feature_wallet_impl/data/repository/CoinPriceRepositoryImpl.kt

```diff
@@ -0,0 +1,100 @@
+    private val remoteCoinPriceDataSource: CoinPriceRemoteDataSource
+) : CoinPriceRepository {
+
+    override suspend fun getCoinPriceAtTime(priceId: String, currency: Currency, timestamp: Long): CoinRate? {
+        var coinRate = cacheCoinPriceDataSource.getFloorCoinPriceAtTime(priceId, currency, timestamp)
+        val hasCeilingItem = cacheCoinPriceDataSource.hasCeilingCoinPriceAtTime(priceId, currency, timestamp)
+        if (coinRate == null && !hasCeilingItem) {
+            val timeInMillis = timestamp.seconds.inWholeMilliseconds
+            val coinRateForAllTime = loadAndCacheForAllTime(priceId, currency)
+            val index = coinRateForAllTime.binarySearchFloor { it.timestamp.compareTo(timeInMillis) }
+            coinRate = coinRateForAllTime.getOrNull(index)
+
+            // If nearest coin rate timestamp is bigger than target timestamp it means that coingecko doesn't have data before coin rate timestamp
+            // so in this case we should return null
+            if (coinRate != null && coinRate.timestamp > timestamp) {
+                return null
+            }
+        }
+
+        return coinRate
+    }
+
+    override suspend fun getCoinPriceRange(priceId: String, currency: Currency, fromTimestamp: Long, toTimestamp: Long): List<HistoricalCoinRate> {
+        // To avoid missing data in case when 'from' and 'to' are the same value,
+        // we need to get data from 1 day before and after
+        val (offsetFrom, offsetTo) = getOffsetFor(fromTimestamp, toTimestamp)
+
+        var coinRates = cacheCoinPriceDataSource.getCoinPriceRange(priceId, currency, offsetFrom, offsetTo)
+        val hasCeilingItem = cacheCoinPriceDataSource.hasCeilingCoinPriceAtTime(priceId, currency, offsetTo)
+        val shouldUpdateCache = coinRates.isEmpty() && !hasCeilingItem
+        if (shouldUpdateCache || timestampIsHigherThanCoinRate(coinRates.lastOrNull(), toTimestamp)) {
+            val timeRange = offsetFrom..offsetTo
+            coinRates = loadAndCacheForAllTime(priceId, currency).filter { it.timestamp in timeRange }
+        }
+
+        return coinRates
+    }
+
+    override suspend fun getCoinRates(priceIds: Set<String>, currency: Currency): Map<String, CoinRateChange?> {
+        return remoteCoinPriceDataSource.getCoinRates(priceIds, currency)
+    }
+
+    override suspend fun getCoinRate(priceId: String, currency: Currency): CoinRateChange? {
+        return remoteCoinPriceDataSource.getCoinRate(priceId, currency)
+    }
+
+    override fun findNearestCoinRate(coinRates: List<HistoricalCoinRate>, timestamp: Long): HistoricalCoinRate? {
+        if (coinRates.isEmpty()) return null
+        if (coinRates.first().timestamp > timestamp) return null // To support the case when the token started trading later than the desired coin rate
+
+        val index = coinRates.binarySearchFloor { it.timestamp.compareTo(timestamp) }
+        return coinRates.getOrNull(index)
+    }
```

Comment:
I beleive this method should not be in CoinPriceRepository but rather be an extension to `List<HistoricalCoinRate>` or an extension to value class wrapping `List<HistoricalCoinRate>`